python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,471 @@
+
+    ## Return this return code if the process must be killed because of timeout
+    TIMEOUT = -999999
+
+    def __init__(self, harness, params):
+        MooseObject.__init__(self, harness, params)
+
+        self.specs = params
+
+        ## The test harness to run callbacks on
+        self.harness = harness
+
+        # Retrieve and store the TestHarness options for use in this object
+        self.options = harness.getOptions()
+
+        # Set max jobs to allow
+        params['max_processes'] = self.options.jobs
+        params['average_load'] = self.options.load
+
+        # For backwards compatibitliy the Scheduler class can be initialized
+        # with no "max_processes" argument and it'll default to a soft limit.
+        # If however a max_processes  is passed we'll treat it as a hard limit.
+        # The difference is whether or not we allow single jobs to exceed
+        # the number of slots.
+        if params['max_processes'] == None:
+            self.job_slots = 1
+            self.soft_limit = True
+        else:
+            self.job_slots = params['max_processes'] # hard limit
+            self.soft_limit = False
+
+        # Requested average load level to stay below
+        self.average_load = params['average_load']
+
+        # Initialize runner pool based on available slots
+        self.runner_pool = ThreadPool(processes=self.job_slots)
+
+        # Initialize status pool to only use 1 process (to prevent status messages from getting clobbered)
+        self.status_pool = ThreadPool(processes=1)
+
+        # Set the time threshold when to report long running jobs
+        self.default_reported_time = 10.0
+
+        # Thread Locking
+        self.thread_lock = threading.RLock()
+
+        # Initialize Scheduler queue objects
+        self.clearAndInitializeJobs()
+
+    ## Clear and Initialize the scheduler queue
+    def clearAndInitializeJobs(self):
+        # Workers in use
+        self.workers_in_use = 0
+
+        # Dictionary of dag testers
+        self.shared_testers_dag = {}
+
+        # A test name to object dictionary containing all tests that are scheduled for execution
+        self.scheduled_name_to_object = {}
+
+        # A set of tests that are actively running
+        self.active_jobs = set([])
+
+        # A tester key to testers objects container {tester : testers}. Used when dealing with dependencies
+        self.scheduled_groups = {}
+
+        # A set holding reported jobs that have already informed the user of a 'RUNNING' status. Tests put in this set
+        # will have the 'FINISHED' caveat applied to them
+        self.reported_jobs = set([])
+
+        # Runner Queue (we put jobs we want to run asyncronously in this)
+        # runner_queue.put( (tester) )
+        self.runner_queue = Queue()
+
+        # Status queue (we put jobs we want to display a status for in this)
+        # status_queue.put( (tester, clock(), subprocess_object) )
+        self.status_queue = Queue()
```

Comment:
Not sure why this comment is labeled as addressed. But I removed the "clear stuff" method all together.